### ISSUE TITLE
[VRL] add support to parse well-known ports to `parse_url()`

### ISSUE THREAD

**Issue by hhromic** [vectordotdev/vector#9374]
### Current Vector Version

```
vector 0.16.1 (x86_64-unknown-linux-gnu 739e878 2021-08-26)
```

### Use-cases

Parsing URL strings into their formal components (scheme, host, port, query, etc) for further individual processing.

### Attempted Solutions

Vector provides the very handy [`parse_url()`](https://vector.dev/docs/reference/vrl/functions/#parse_url) function for parsing the components of a URL string:
```coffee
parse_url!(s'https://dict.youdao.com:443/fsearch?keyfrom=deskdict#frag1%22)')
# { "fragment": "frag1%22)", "host": "dict.youdao.com", "password": "", "path": "/fsearch", "port": null, "query": { "keyfrom": "deskdict" }, "scheme": "https", "username": "" }
```

However, as can be noted in the above example, the `port` field sometimes can be `null`. This is because Vector is using the `url` standard library for parsing: https://docs.rs/url/2.2.2/url/struct.Url.html

In particular, Vector uses the [`port()`](https://docs.rs/url/2.2.2/url/struct.Url.html#method.port
) instance method to obtain the port field:

https://github.com/vectordotdev/vector/blob/932552ea24c3a2fe887abf713ce512b3a5b1c0c3/lib/vrl/stdlib/src/parse_url.rs#L86

The `port()` method has the following note explaining the observed behaviour:

> Note that default port numbers are never reflected by the serialization, use the port_or_known_default() method if you want a default port number returned.

Indeed Vector inherits this behaviour (does not parse well-known ports):
```coffee
parse_url!(s'http://dict.youdao.com:79/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": 79, "query": {  }, "scheme": "http", "username": "" }
parse_url!(s'http://dict.youdao.com:80/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": null, "query": {  }, "scheme": "http", "username": "" }
parse_url!(s'http://dict.youdao.com/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": null, "query": {  }, "scheme": "http", "username": "" }
parse_url!(s'http://dict.youdao.com:81/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": 81, "query": {  }, "scheme": "http", "username": "" }
```
```coffee
parse_url!(s'wss://dict.youdao.com:442/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": 442, "query": {  }, "scheme": "wss", "username": "" }
parse_url!(s'wss://dict.youdao.com/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": null, "query": {  }, "scheme": "wss", "username": "" }
parse_url!(s'wss://dict.youdao.com:443/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": null, "query": {  }, "scheme": "wss", "username": "" }
parse_url!(s'wss://dict.youdao.com:444/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": 444, "query": {  }, "scheme": "wss", "username": "" }
```

Unfortunately this behaviour is a bit counter-intuitive as one would expect the `port` field to be always populated except when truly a port cannot be derived from a URL string.

At the moment this behaviour is forcing us to write workaround code such as this:
```coffee
.port = if ( .port == null && .scheme == "http" ) { 80 }
.port = if ( .port == null && .scheme == "https" ) { 443 }
.port = if ( .port == null && .scheme == "ws" ) { 80 }
.port = if ( .port == null && .scheme == "wws" ) { 443 }
.port = if ( .port == null && .scheme == "ftp" ) { 21 }
```

### Proposal

Following the advice from the `url` library note for `port()`, there is also the [`port_or_known_default()`](https://docs.rs/url/2.2.2/url/struct.Url.html#method.port_or_known_default) method:

> Return the port number for this URL, or the default port number if it is known.
> This method only knows the default port number of the http, https, ws, wss and ftp schemes.

Essentially implementing the workaround code in the previous section.

Therefore a natural solution proposal is to simply replace the call to `port()` with a call to `port_or_known_default()` in `parse_url()` implementation in Vector.

Of course, the above is a breaking change. An alternative could then be to implement an extra parameter in `parse_url()` to specify the type of port-parsing desired. For example `parse_known_ports: "true"`. However, I could guess that most users would prefer to have `port` populated than the current behaviour so not sure if it's worth it.

**Comment by jszwedko** [comment]
Thanks for this @hhromic . I agree with you, even though it would technically be a breaking change, that returning the default port for the scheme makes sense. Especially in the:

```
$ parse_url!(s'http://dict.youdao.com:80/')
# { "fragment": null, "host": "dict.youdao.com", "password": "", "path": "/", "port": null, "query": {  }, "scheme": "http", "username": "" }
```

case.

**Comment by hhromic** [comment]
@jszwedko yes the case where an explicit well-known port still returns `null` was particularly annoying :)

Thanks for considering the feature request, much appreciated!

**Comment by jszwedko** [comment]
👍 

https://github.com/servo/rust-url/issues/706 seems to be related for the "not retrieving normalized port" case.

**Comment by hhromic** [comment]
I'm not sure we need to be concerned about that issue. If I understood it correctly, is about the case of schemes with default ports that are not _well-known_ ports?  I don't think Vector falls into an application that needs to deal with that.

Unless I'm not understanding the issue very well :(

**Comment by jszwedko** [comment]
My understanding of that issue is that they want to _require_ the user to specify a port, even if the scheme is `http` and currently it isn't possible to tell if the user specified a port of that port is the default for the scheme.

After thinking about this a bit more, I'm rethinking whether returning `port_or_known_default()` is best as it obfuscates whether the input actually contains a port or not. My current thinking is we:

* Modify `parse_url` to return the port, if in the input, even if it is the default port. This solves the `parse_url!(s'http://dict.youdao.com:80/')` case.
* Add a `default_known_ports` parameter that can be used to default the port if the scheme is known

I agree with you that Vector users are likely to, typically, want the port even if it is the default though; so we can discuss if `default_known_ports` should be `true` or not.

How does that sound to you?

**Comment by hhromic** [comment]
> * Modify `parse_url` to return the port, if in the input, even if it is the default port. This solves the `parse_url!(s'http://dict.youdao.com:80/')` case.

Umm I think is better to delegate the functionality to the URL crate as much as possible and avoid Vector assuming a workaround that needs maintenance. In other words you can keep saying "Vector just gives you access to what the URL crate does".

After all, the documentation is crystal clear and that is what matters in any API:
> Note that default port numbers are never reflected by the serialization, use the port_or_known_default() method if you want a default port number returned.

There must be a reason for that choice (I would like to know why hehe).
Maybe that note can be added to the `parse_url()` documentation regarding the new boolean option to be super clear of what it does?

I saw the workaround you had to do circumvent the crate's "limitation" and I think is not necessary really imho.

> * Add a `default_known_ports` parameter that can be used to default the port if the scheme is known

This would give the Vector user a choice indeed and I think is the best solution for the dilemma. I'm very confident that users will most of the time want the new behaviour of always having a port number parsed, so defaulting this to `true` is very senseful to me.

In other words the boolean just switches which port-parsing function to use: `port()` vs `port_or_known_defaults()`. And both are well-defined in their upstream documentations.

**Comment by hhromic** [comment]
> My understanding of that issue is that they want to _require_ the user to specify a port, even if the scheme is `http` and currently it isn't possible to tell if the user specified a port of that port is the default for the scheme.

I don't think any user of Vector really would care about that. At the end of the day the important thing is to know which port the URL made reference to. I highly doubt that any observability data being processed by Vector requires to consider non-well-known ports for these schemes.

**Comment by jszwedko** [comment]
@hhromic Thanks for the thoughts and responses!

> > * Modify `parse_url` to return the port, if in the input, even if it is the default port. This solves the `parse_url!(s'http://dict.youdao.com:80/')` case.
> 
> Umm I think is better to delegate the functionality to the URL crate as much as possible and avoid Vector assuming a workaround that needs maintenance. In other words you can keep saying "Vector just gives you access to what the URL crate does".

In this case, I think what the behavior of the `url` crate is not exactly right for Vector's use-case where we expose URL parsing functionality. The crate doesn't strike me as designed for data processing tasks so I think it's ok if we need to layer on a little bit to provide the best experience to Vector users.
 
> After all, the documentation is crystal clear and that is what matters in any API:
> 
> > Note that default port numbers are never reflected by the serialization, use the port_or_known_default() method if you want a default port number returned.
> There must be a reason for that choice (I would like to know why hehe). Maybe that same note can be added to the `parse_url()` documentation regarding the new boolean option to be super clear of what it does?

The documentation is clear, but I still think it is surprising. I'm surprised that crate has any protocol-aware functionality. For example, it defaults `http', `ftp`, `ws`, `wss`, `ftp`, but not `ftps`, `ssh`, `irc`, etc.. While I concede the ones they've chosen to default (perhaps minus `ftp`) or the most common schemes Vector users are likely to see in their logs, the decision of the `url` crate on which to default seems rather arbitrary.

> I saw the workaround you had to do to circumvent the crate's "limitation" and I think is not necessary really imho.
> 
> > * Add a `default_known_ports` parameter that can be used to default the port if the scheme is known
> 
> This would give the Vector user a choice indeed and I think is the best solution for the dilemma. I'm very confident that users will most of the time want the new behaviour of always having a port number parsed, so defaulting this to `true` is very senseful to me.
> 
> In other words the boolean just switches which port-parsing function to use: `port()` vs `port_or_known_defaults()`. And both are well-defined in their upstream documentations.

I can see an argument for defaulting to the behavior of parsing the port for the "known" schemes. I can also, however, see the argument that it should return the port only if it was actually in the URL. Defaulting to not returning default ports would also be less of an incompatible change.

> > My understanding of that issue is that they want to _require_ the user to specify a port, even if the scheme is `http` and currently it isn't possible to tell if the user specified a port of that port is the default for the scheme.
> 
> I don't think any user of Vector really would care about that. At the end of the day the important thing is to know which port the URL made reference to. I highly doubt that any observability data being processed by Vector requires to consider non-well-known ports for these schemes.

Right, but I think it is more up to the user-agent interpreting the URL to decide what port to use by default given the scheme than a URL parsing library.

**Comment by hhromic** [comment]
Hi @jszwedko, those are good points. Perhaps I was -unintentionally- becoming too opinionated on this one.

> In this case, I think what the behavior of the `url` crate is not exactly right for Vector's use-case where we expose URL parsing functionality. The crate doesn't strike me as designed for data processing tasks so I think it's ok if we need to layer on a little bit to provide the best experience to Vector users.

I guess is ok for me too. As long as Vector is well-documented on what it does and how, everyone should be happy.

> The documentation is clear, but I still think it is surprising. I'm surprised that crate has any protocol-aware functionality. For example, it defaults `http'` `ftp`, `ws`, `wss`, `ftp`, but not `ftps`, `ssh`, `irc`, etc.. While I concede the ones they've chosen to default (perhaps minus `ftp`) are the most common schemes Vector users are likely to see in their logs, the decision of the `url` crate on which to default seems rather arbitrary.

Yes indeed their selection of "well-known" scheme ports is a bit inconsistent, specially having `ftp` but lacking `ftps`.

> I can see an argument for defaulting to the behavior of parsing the port for the "known" schemes. I can also, however, see the argument that it should return the port only if it was actually in the URL. Defaulting to not returning default ports would also be less of an incompatible change.

How about we leave it as in your current PR? Namely:
* Always return port numbers if explicitly given in the input URL
* There is an option to return well-known ports or not when they are not given in the URL

What do you think? Hope this can be ready for the next version of Vector.

**Comment by jszwedko** [comment]
Thanks for taking another look @hhromic ! I agree with your conclusion there. That is:

* Always return port numbers if explicitly given in the input URL
* There is an option to return well-known ports or not when they are not given in the URL

I think this can make it into 0.17.0 which'll be released next week.

**Comment by jszwedko** [comment]
Uggggh. 

Having read over https://url.spec.whatwg.org/#concept-basic-url-parser it does seem like URL parsing is well-defined as:

> Set url’s port to null, if port is url’s scheme’s default port, and to port otherwise.

And the schemes the `url` crate is handling are actually documented in the spec as ["special schemes"](https://url.spec.whatwg.org/#special-scheme)

> scheme | port
> -- | --
> "ftp" | 21
> "file" | null
> "http" | 80
> "https" | 443
> "ws" | 80
> "wss" | 443

Given this new information, I think we should just implement the spec behavior of setting the port to `null` (as surprising as this was to me) if it is set to the default port. Subsequently, we add the optional parameter to return the well-known ports.

**Comment by hhromic** [comment]
> Uggggh.
> 
> Having read over https://url.spec.whatwg.org/#concept-basic-url-parser it does seem like URL parsing is well-defined as:
> 
> > Set url’s port to null, if port is url’s scheme’s default port, and to port otherwise.
> 
> And the schemes the `url` crate is handling are actually documented in the spec as ["special schemes"](https://url.spec.whatwg.org/#special-scheme)
> 
> > scheme
> > port
> > 
> > 
> > 
> > 
> > "ftp"
> > 21
> > 
> > 
> > "file"
> > null
> > 
> > 
> > "http"
> > 80
> > 
> > 
> > "https"
> > 443
> > 
> > 
> > "ws"
> > 80
> > 
> > 
> > "wss"
> > 443
> 

Very interesting finding (for me). I wasn't aware of WHATWG, only mostly RFCs and the W3C.

> Given this new information, I think we should just implement the spec behavior of setting the port to `null` (as surprising as this was to me) if it is set to the default port. Subsequently, we add the optional parameter to return the well-known ports.

Yes I fully agree with this now.

So in summary, default behaviour `default_known_ports=false` in `parse_url()` will be like today:
* `http://some.domain` -> `port = null`
* `http://some.domain:80` -> `port = null`
* `https://some.domain:445` -> `port = 445`

And with `default_known_ports=true` it will be:
* `http://some.domain` -> `port = 80`
* `http://some.domain:80` -> `port = 80`
* `https://some.domain:445` -> `port = 445`

This works perfectly for us! Thank you very much for the interesting discussion and implementing the functionality.
